modflow_devtools/netcdf_schema.py

```diff
@@ -0,0 +1,206 @@
+    if not path.is_file():
+        raise AssertionError(f"Not a valid mf6 component: {toml_name}")
+    with path.open(mode="rb") as toml_file:
+        return Dfn.load(toml_file, name=toml_name, version=2)
+
+
+def validate(v, dims: list[int]):
+    """
+    Validate model NetCDF specification against schema.
+
+    Parameters
+    ----------
+    dims: list
+        modeflow 6 model grid dimensions
+        [nlay, nrow, ncol] or [nlay, ncpl]
+    """
+    try:
+        NetCDFModel.model_validate(v, context={"dims": dims})
+    except ValidationError:
+        raise
+
+
+class NetCDFModel(BaseModel):
+    attrs: "NetCDFModelAttrs"
+    variables: list["NetCDFPackageParam"] = Field(default_factory=list)
+
+    @field_validator("attrs", mode="before")
+    @classmethod
+    def validate_attrs(cls, v: Any) -> Any:
+        """
+        validate model (dataset) scoped attributes dictionary
+        """
+        v = {k.lower(): v.lower() if isinstance(v, str) else v for k, v in v.items()}
+        return v
+
+
+class NetCDFModelAttrs(BaseModel):
+    # order of params dictates when data added to info dict
+    mesh: str | None = Field(default=None)
+    modflow_grid: str = Field()
+    modflow_model: str = Field()
+
+    @field_validator("mesh", mode="before")
+    @classmethod
+    def validate_mesh(cls, v: Any, info: ValidationInfo) -> Any:
+        if v is not None:
+            assert v == "layered"
+            info.context["mesh"] = v  # type: ignore
+        return v
+
+    @field_validator("modflow_grid", mode="before")
+    @classmethod
+    def validate_modflow_grid(cls, v: Any, info: ValidationInfo) -> Any:
+        mesh = info.data.get("mesh")
+        dims = info.context.get("dims")  # type: ignore
+        if mesh is None:
+            assert v == "structured"
+            if len(dims) != 3:
+                raise AssertionError(f"Expected 3 grid dimensions: {dims}")
+        else:
+            if len(dims) != 2:
+                raise AssertionError(
+                    f"Expected 2 grid dimensions for layered mesh: {dims}"
+                )
+        info.context["grid"] = v  # type: ignore
+        return v
+
+    @field_validator("modflow_model", mode="before")
+    @classmethod
+    def validate_modflow_model(cls, v: Any, info: ValidationInfo) -> Any:
+        tokens = v.split(":")
+        if len(tokens) != 2:
+            raise ValueError(f"Invalid modflow_model attribute: {v}")
+        modeltype = tokens[0].strip()
+        if modeltype[-1].isdigit():
+            modeltype = modeltype[:-1]
+        info.context["modeltype"] = modeltype  # type: ignore
+        info.context["modelname"] = tokens[1].strip()  # type: ignore
+        return v
+
+
+class NetCDFPackageParam(BaseModel):
+    param: str = Field()
+    shape: list[str] = Field(default_factory=list)
+    attrs: "NetCDFParamAttrs"
+    encodings: "NetCDFParamEncodings"
+    varname: str = Field()
+    numeric_type: str = Field()
```

Comment:
does `numeric_type` correspond directly to the types enumerated [here](docs.unidata.ucar.edu/nug/current/md_types.html)? if so, suggest renaming `dtype` as it's not necessarily numeric (e.g. char) and `dtype` is conventional in numpy/pandas/xarray/etc

Reply:
done in https://github.com/modflowpy/pyphoenix-project/pull/288